core/executor.py

```diff
@@ -632,6 +639,62 @@ def CaptureStdout(self, node):
 
         return status, stdout_str
 
+    def CaptureOutputs(self, node):
+        # type: (command_t) -> Tuple[int, str, str]
+
+        p = self._MakeProcess(node, self.exec_opts.inherit_errexit(),
+                              self.exec_opts.errtrace())
+        # Shell quirk: Command subs remain part of the shell's process group, so we
+        # don't use p.AddStateChange(process.SetPgid(...))
+
+        stdout_fd, w = posix.pipe()
+        stderr_fd, w2 = posix.pipe()
+        p.AddStateChange(process.StdoutToPipe(stdout_fd, w))
+        p.AddStateChange(process.StderrToPipe(stderr_fd, w2))
+
+        p.StartProcess(trace.CommandSub)
+        #log('Command sub started %d', pid)
+
+        stdout_chunks = []  # type: List[str]
+        stderr_chunks = []  # type: List[str]
+        posix.close(w)  # not going to write
+        posix.close(w2)  # not going to write
+        open_fds = [stdout_fd, stderr_fd]
+        while True:
+            fds = pyos.WaitForReading(open_fds)
+
+            # zero outputs mean something went wrong
+            if len(fds) == 0:
+                break
+
+            for fd in fds:
+                if fd == stdout_fd:
+                    n, err_num = pyos.Read(fd, 4096, stdout_chunks)
+                else:
+                    n, err_num = pyos.Read(fd, 4096, stderr_chunks)
+                if n < 0:
+                    if err_num == EINTR:
+                        pass  # retry
+                    else:
+                        # Like the top level IOError handler
+                        e_die_status(
+                            2,
+                            'Oils I/O error (read): %s' % posix.strerror(err_num))
+                elif n == 0:  # EOF
+                    open_fds.remove(fd)
+
+            if len(open_fds) == 0:
```

Comment:
I am not sure we need both `break` conditions ... but I guess it avoids calling select() with an empty list, so maybe that's useful

Reply:
According to [stackoverflow](https://stackoverflow.com/questions/15109527/return-value-of-select-with-empty-sets):

>     Some code calls select() with all three sets empty, nfds zero, and a non-NULL timeout as a fairly portable way to sleep with subsecond precision.

I have the feeling we might get an indefinite sleep without the second break...

Reply:
Ah OK good point, thanks for looking that up